AHS plugin documentation (#150)
aws/amazon-braket-pennylane-plugin-python: doc/devices/ahs_local.rst

The local AHS device
====================

The local AHS device of the PennyLane-Braket plugin runs Analogue Hamiltonian Simulation on the local Braket SDK. This
could be either utilizing the processors of your own PC, or those of a `Braket notebook instance <https://docs.aws.amazon.com/braket/latest/developerguide/braket-get-started-create-notebook.html>`_ hosted on AWS.

This device is useful for small-scale simulations in which the time of sending a job to a remote service would add
an unnecessary overhead. It can also be used for rapid prototyping before running a computation
on a paid-for remote service.

Usage
~~~~~

> After the Braket SDK and the plugin are installed you immediately have access to the `local Braket AHS device <https://docs.aws.amazon.com/braket/latest/developerguide/braket-devices.html#braket-simulator-ahs-local>`_ in PennyLane.

Reviewer: The link is pointing to the "local AHS simulator", rather than a device (in the PennyLane nomenclature), so I'd suggest to say
```suggestion
After the Braket SDK and the plugin are installed you immediately have access to the `local Braket AHS simulator <https://docs.aws.amazon.com/braket/latest/developerguide/braket-devices.html#braket-simulator-ahs-local>`_ in PennyLane.
```

> After the Braket SDK and the plugin are installed you immediately have access to the `local Braket AHS device <https://docs.aws.amazon.com/braket/latest/developerguide/braket-devices.html#braket-simulator-ahs-local>`_ in PennyLane.

Reviewer: Should we mention that `jax` and `jaxlib` (`0.4.3`) need to be installed?

Author: Added!

Reviewer 2: So technically for the local and remote ahs devices we shouldnt need jax, since jax is only required for the numerical ode solver. So potentially we should change when and where that import error is raised in `ParametrizedEvolution` 🤔

> After the Braket SDK and the plugin are installed you immediately have access to the `local Braket AHS device <https://docs.aws.amazon.com/braket/latest/developerguide/braket-devices.html#braket-simulator-ahs-local>`_ in PennyLane.

Reviewer 2: I mean in `BraketAwsAhsDevice` and `BraketLocalAhsDevice`. Something like a note that all parameters get multiplied by `10**6`

> After the Braket SDK and the plugin are installed you immediately have access to the `local Braket AHS device <https://docs.aws.amazon.com/braket/latest/developerguide/braket-devices.html#braket-simulator-ahs-local>`_ in PennyLane.

Author: I believe it's already noted in all the functions where units are relevant. Sometimes in in-line comments in the code, but since almost all the functions are private and not user-facing, I think that's fine (even preferable, as it keeps the comment close to where its used). They're all private methods.

> After the Braket SDK and the plugin are installed you immediately have access to the `local Braket AHS device <https://docs.aws.amazon.com/braket/latest/developerguide/braket-devices.html#braket-simulator-ahs-local>`_ in PennyLane.

Reviewer 2: You mean the functions in PennyLane, right? Since in PennyLane itself no conversions are happening and this is something specific to the device, we should make it clear in the device docs as well. Something like a `note` block on the device docs
![image](https://user-images.githubusercontent.com/43949391/233330434-c2a00f43-ffc9-4d08-bb38-0091718957cb.png)

> After the Braket SDK and the plugin are installed you immediately have access to the `local Braket AHS device <https://docs.aws.amazon.com/braket/latest/developerguide/braket-devices.html#braket-simulator-ahs-local>`_ in PennyLane.

Author: > Would be good to link to Braket SDK, ideally with install instructions

Added!

The local AHS device is not gate-based. Instead, it is compatible with the ``ParametrizedEvolution`` operator from `pulse programming <https://docs.pennylane.ai/en/stable/code/qml_pulse.html>`_ in PennyLane.

To instantiate the local Braket simulator, simply use:

.. code-block:: python

    import pennylane as qml
    device_local = qml.device("braket.local.ahs", wires=2)

This device can be used with a QNode within PennyLane. It accepts circuits with a single ``ParametrizedEvolution``

> operator based on a ``ParametrizedHamiltonian`` compatible with the simulated hardware. More information about creating

Reviewer: It was not entirely clear to me how "the simulated hardware" was specified/chosen at this point. Is there just a single one when setting up the PL device above?

Author: Yes, there's only one simulator

> operator based on a ``ParametrizedHamiltonian`` compatible with the simulated hardware. More information about creating

Reviewer 2: Do hyper-links work on plugin websites? If not, could do regular links to the respective docs

Author: You don't think the link at the end of the paragraph to the pulse section of the PennyLane docs is enough? Making them regular links will change the formatting from the "code" appearance to "link" appearance.

Reviewer 2: If the hyper links `class:'~.ParametrizedHamiltonian'` dont work then yeah I'd prefer regular links over formatting

Author: Done!

PennyLane operators for AHS can be found in the `PennyLane docs <https://docs.pennylane.ai/en/stable/code/qml_pulse.html>`_.

Creating a register
^^^^^^^^^^^^^^^^^^^

The atom register defines where the atoms will be located, which determines the strength of the interaction
between the atoms. Here we define coordinates for the atoms to be placed at (in micrometers), and create a constant
interaction term for the Hamiltonian:

.. code-block:: python

>     coordinates = [[0, 0], [0, 5]]  # number of coordinate pairs must match number of device wires

Reviewer: This comment leads to a horizontal scroll bar, maybe it's nicer to do
```suggestion
    # number of coordinate pairs must match number of device wires
    coordinates = [[0, 0], [0, 5]]  
```

We can create a drive with a global component and (positive) local detunings. If the local detunings are time-dependent,
they must all have the same time-dependent envelope, but can have different, positive scaling factors.

.. code-block:: python

    # gaussian amplitude function (qml.pulse.rect enforces 0 at start and end for hardware)
    def amp_fn(p, t):
        f = p[0] * jnp.exp(-(t-p[1])**2/(2*p[2]**2))
        return qml.pulse.rect(f, windows=[0.1, 1.7])(p, t)

    # defining a linear detuning
    def det_fn_global(p, t):
        return p * t

    def det_fn_local(p, t):
        return p * t**2

    # creating a global drive on all wires
    H_global = qml.pulse.rydberg_drive(amplitude=amp_fn, phase=0, detuning=det_fn_global, wires=[0, 1])

    # creating local drives
    H_local0 = qml.pulse.rydberg_drive(amplitude=0, phase=0, detuning = det_fn_local, wires=[0])
    H_local1 = qml.pulse.rydberg_drive(amplitude=0, phase=0, detuning = det_fn_local, wires=[1])

    # full hamiltonian
    H = H_interaction + H_global + H_local0 + H_local1

Executing an AHS program
^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: python

    @qml.qnode(device_local)
    def circuit(params):

>         qml.evolve(H)(params, t)

Reviewer: This line raises an error when executing the program because `t` has not been defined.
When providing some time (`t=2.`), I received the error
```
NotImplementedError: Multiple pulses in a Hamiltonian are not currently supported. Received 3 pulses.
```
Probably I checked out the wrong version of the plugin repository? :thinking:

Author: Ah, thanks for the heads up, I'll correct that.

Yes, that part will only work on the `rh_shift` [branch](https://github.com/lillian542/amazon-braket-pennylane-plugin-python/pull/7)

        return qml.sample()

    # amp_fn expects p to contain 3 parameters
    amp_params = [2.5, 1, 0.3]
    # global_det_fn expects p to be a single parameter
    det_global_params = 0.2
    # each of the local drives take a single parameter for p
    local_params1 = 0.5

>     local_params2 = 1

Author: In terms of a comment in the main text, I feel like this is covered by the introductory sentence in the drive section: "We can create a drive with a global component and (positive) local detunings. If the local detunings are time-dependent, they must all have the same time-dependent envelope, but can have different, positive scaling factors."

I added a clarifying comment above the line where the local detunings are created, and above this line. Let me know what you think!

>     local_params2 = 1

Reviewer 2: looks good 👍

We can create a drive with a global component and (positive) local detunings. If the local detunings are time-dependent,
they must all have the same time-dependent envelope, but can have different, positive scaling factors.

.. code-block:: python

    # gaussian amplitude function (qml.pulse.rect enforces 0 at start and end for hardware)
    def amp_fn(p, t):
        f = p[0] * jnp.exp(-(t-p[1])**2/(2*p[2]**2))
        return qml.pulse.rect(f, windows=[0.1, 1.7])(p, t)

    # defining a linear detuning
    def det_fn_global(p, t):
        return p * t

    def det_fn_local(p, t):
        return p * t**2

    # creating a global drive on all wires
    H_global = qml.pulse.rydberg_drive(amplitude=amp_fn, phase=0, detuning=det_fn_global, wires=[0, 1])

    # creating local drives
    H_local0 = qml.pulse.rydberg_drive(amplitude=0, phase=0, detuning = det_fn_local, wires=[0])
    H_local1 = qml.pulse.rydberg_drive(amplitude=0, phase=0, detuning = det_fn_local, wires=[1])

    # full hamiltonian
    H = H_interaction + H_global + H_local0 + H_local1

Executing an AHS program
^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: python

    @qml.qnode(device_local)
    def circuit(params):

>         qml.evolve(H)(params, t)

Author: ```suggestion
        qml.evolve(H)(params, t=1.5)
```

The local AHS device
====================

> The local analog hamiltonian simulation (AHS) device of the PennyLane-Braket plugin runs Analogue Hamiltonian Simulation on the local Braket SDK. This

Reviewer 3: ```suggestion
The local analog Hamiltonian simulation (AHS) device of the PennyLane-Braket plugin runs Analogue Hamiltonian Simulation on the local Braket SDK. This
```

could be either utilizing the processors of your own PC, or those of a `Braket notebook instance <https://docs.aws.amazon.com/braket/latest/developerguide/braket-get-started-create-notebook.html>`_ hosted on AWS.

This device is useful for small-scale simulations in which the time of sending a job to a remote service would add
an unnecessary overhead. It can also be used for rapid prototyping before running a computation
on a paid-for remote service.

Usage
~~~~~

After the Braket SDK and the plugin are installed you immediately have access to the `local Braket AHS simulator <https://docs.aws.amazon.com/braket/latest/developerguide/braket-devices.html#braket-simulator-ahs-local>`_ in PennyLane.

The local AHS device is not gate-based. Instead, it is compatible with the `ParametrizedEvolution <https://docs.pennylane.ai/en/stable/code/api/pennylane.pulse.ParametrizedEvolution.html>`_
operator from `pulse programming <https://docs.pennylane.ai/en/stable/code/qml_pulse.html>`_ in PennyLane.

> Note that pulse programming in PennyLane requires the module ``jax``. You can install jax via: pip install jax==0.4.3 jaxlib==0.4.3

Reviewer 3: ```suggestion
Note that pulse programming in PennyLane requires the module ``jax``. You can install jax via:

>>> pip install jax==0.4.3 jaxlib==0.4.3
```

To instantiate the local Braket simulator, simply use:

.. code-block:: python

    import pennylane as qml
    device_local = qml.device("braket.local.ahs", wires=2)

This device can be used with a QNode within PennyLane. It accepts circuits with a single `ParametrizedEvolution <https://docs.pennylane.ai/en/stable/code/api/pennylane.pulse.ParametrizedEvolution.html>`_
operator based on a `ParametrizedHamiltonian <https://docs.pennylane.ai/en/stable/code/api/pennylane.pulse.ParametrizedHamiltonian.html>`_ compatible with the simulated hardware.
More information about creating PennyLane operators for AHS can be found in the `PennyLane docs <https://docs.pennylane.ai/en/stable/code/qml_pulse.html>`_.

.. note::
    It is important to keep track of units when specifying electromagnetic pulses for hardware control.
    The frequency and amplitude provided in PennyLane will be multiplied to a factor of 1e6 when translating

>     to hardware (converted from :math:`2 \pi` MHz to rad/s), while time will be divided by 1e6 (converted from

Author: Kind of. Currently the plan is to ask PL users to specify in units of $2 \pi$ MHz. But also the plan is to settle this for sure Wednesday morning when Korbinian is back.

>     to hardware (converted from :math:`2 \pi` MHz to rad/s), while time will be divided by 1e6 (converted from

Reviewer 2: The idea is to have as little implicitly assumed factors as possible, such that the rule "what the users puts, is also what the program uses" holds. 

Currently, the only caveat is that we still multiply/divide by `1e6` on the device. This is because in numerical simulation these factors cancel out so it makes sense to not put them explicitly, but on the real device it is expecting values in the actual physical order of magnitude.

Alternative is to not have this multiplication and tell users that they need to specify this when running on real hardware. It would make the user more autonomous, but the downside is that you have to manually do this step when going from simulation to real device.

>     to hardware (converted from :math:`2 \pi` MHz to rad/s), while time will be divided by 1e6 (converted from

Author: Settled in meeting today - we are expecting standard frequency in PL and converting to angular frequency for the user. We are also converting orders of magnitude so that the order of magnitude expected by the PL ODE will be converted to the correct units for users, rather than asking users to rescale all their parameters to go from PL to Braket.

>     to hardware (converted from :math:`2 \pi` MHz to rad/s), while time will be divided by 1e6 (converted from

Author: So, none of these parameters will be converted. They will be passed into functions defining amplitude and detuning, and the outputs of _those_ functions will be converted from standard to angular frequency. 

I think we should focus on highlighting expected input units, as well as what units the hardware specs are provided in on the Braket end, and beyond that not try to focus too much on where unit conversion is occurring. If they know what units they are reading and what units they should be inputing, they are good to go, and if they want more detail about specific implementation they can read the code.

I am working on adding some documentation and examples on the PennyLane side today, and I think this plus the demo will be good documentation, and we should keep this as is.

The local AHS device
====================

The local analog Hamiltonian simulation (AHS) device of the PennyLane-Braket plugin runs simulation on the local Braket SDK. This
could be either utilizing the processors of your own PC, or those of a `Braket notebook instance <https://docs.aws.amazon.com/braket/latest/developerguide/braket-get-started-create-notebook.html>`_ hosted on AWS.

This device is useful for small-scale simulations in which the time of sending a job to a remote service would add
an unnecessary overhead. It can also be used for rapid prototyping before running a computation
on a paid-for remote service.

Usage
~~~~~

> Connecting to the device will require installing `Braket SDK <https://github.com/aws/amazon-braket-sdk-python#prerequisites>`_, which can be done via

Reviewer 4: ```suggestion
Connecting to the device will require installing the `Amazon Braket SDK <https://github.com/aws/amazon-braket-sdk-python#prerequisites>`_, which can be done via
```

> Connecting to the device will require installing the `Amazon Braket SDK <https://github.com/aws/amazon-braket-sdk-python#prerequisites>`_, which can be done via
> 
> >>> pip install amazon-braket-sdk
> 

Reviewer 3: ```suggestion
```
It's not clear to me that this needs to be mentioned, and it isn't for the local qubit simulator. 

Also, isn't `amazon-braket-sdk` a [dependency](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/605252299cb5fd7b8417b41b24802ac9f2c9a345/setup.py#L30) of the plugin?

After the Braket SDK and the plugin are installed you immediately have access to the `local Braket AHS simulator <https://docs.aws.amazon.com/braket/latest/developerguide/braket-devices.html#braket-simulator-ahs-local>`_ in PennyLane.

The local AHS device is not gate-based. Instead, it is compatible with the `ParametrizedEvolution <https://docs.pennylane.ai/en/stable/code/api/pennylane.pulse.ParametrizedEvolution.html>`_
operator from `pulse programming <https://docs.pennylane.ai/en/stable/code/qml_pulse.html>`_ in PennyLane.

> Note that pulse programming in PennyLane requires the module ``jax``. You can install jax via:
> 
> >>> pip install jax==0.4.3 jaxlib==0.4.3

Reviewer 3: ```suggestion
Note that pulse programming in PennyLane requires the module ``jax``, which can be installed
following the instructions [here](https://github.com/google/jax#installation).
```

    # defining a linear detuning
    def det_fn_global(p, t):
        return p * t

    def det_fn_local(p, t):
        return p * t**2

    # creating a global drive on all wires
    H_global = qml.pulse.rydberg_drive(amplitude=amp_fn, phase=0, detuning=det_fn_global, wires=[0, 1])

    # creating local drives
    # note only local detuning is currently supported, so amplitude and phase are set to 0
    H_local0 = qml.pulse.rydberg_drive(amplitude=0, phase=0, detuning = det_fn_local, wires=[0])
    H_local1 = qml.pulse.rydberg_drive(amplitude=0, phase=0, detuning = det_fn_local, wires=[1])

    # full hamiltonian
    H = H_interaction + H_global + H_local0 + H_local1

Executing an AHS program
^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: python

    @qml.qnode(device_local)
    def circuit(params):
        qml.evolve(H)(params, t=1.5)
        return qml.sample()

    # amp_fn expects p to contain 3 parameters
    amp_params = [2.5, 1, 0.3]
    # global_det_fn expects p to be a single parameter
    det_global_params = 0.2
    # each of the local drives take a single parameter for p
    # the detunings have the same shape, but vary by scaling factor p
    local_params1 = 0.5

>     local_params2 = 1

Reviewer 2: Can you mention in code comments or text which parameters will be multiplied by `2 pi 1e6`? Its actually not clear to myself right now

Author: These parameters are not multiplied by anything - they are input for the functions that define amplitude and detunings. The amplitudes and detunings output by these functions are converted from MHz to rad/s.

I don't think adding information about this here would be clarifying. I think we have a good plan for documentation and examples on the PennyLane side in the helper functions and demos, and we should leave this as is.

.. note::
    It is important to keep track of units when specifying electromagnetic pulses for hardware control.
    The frequency and amplitude provided in PennyLane for Rydberg atom systems are expected to be in units of MHz,
    time in microseconds, phase in radians, and distance in micrometers. All of these will be converted to SI units
    internally as needed for upload to the hardware, and frequency will be converted to angular frequency
    (multiplied by :math:`2 \pi`).

    When reading hardware specifications from the Braket backend, bear in mind that all units are SI and frequencies
    are in rad/s. This conversion is done when creating a pulse program for upload, and units in the PennyLane

>     functions should follow the conventions specified in the docstrings to ensure correct unit conversion.

Reviewer 2: is "the docstring" referring to _this_ docstring or another one?

Author: good point 😅

Author: I think it's clearer now
